### ISSUE TITLE
[JENKINS-44772] User Scoped credentials are not used by the "withCredentials" pipeline step

### ISSUE THREAD

**Issue by jenkins-infra-bot** [jenkinsci/credentials-binding-plugin#444]
<h1><a name="Environment"></a>Environment</h1>

<ul>
	<li>Jenkins 2.46.2</li>
	<li>credentials:2.1.13 'Credentials Plugin'</li>
	<li>credentials-binding:1.11 'Credentials Binding Plugin'</li>
	<li>workflow-aggregator:2.5 'Pipeline'</li>
	<li>workflow-api:2.12 'Pipeline: API'</li>
	<li>workflow-basic-steps:2.4 'Pipeline: Basic Steps'</li>
	<li>workflow-cps:2.29 'Pipeline: Groovy'</li>
	<li>workflow-cps-checkpoint:2.4 'CloudBees Pipeline: Groovy Checkpoint Plugin'</li>
	<li>workflow-cps-global-lib:2.7 'Pipeline: Shared Groovy Libraries'</li>
	<li>workflow-durable-task-step:2.10 'Pipeline: Nodes and Processes'</li>
	<li>workflow-job:2.10 'Pipeline: Job'</li>
	<li>workflow-multibranch:2.14 'Pipeline: Multibranch'</li>
	<li>workflow-scm-step:2.4 'Pipeline: SCM Step'</li>
	<li>workflow-step-api:2.9 'Pipeline: Step API'</li>
	<li>workflow-support:2.14 'Pipeline: Supporting APIs'</li>
</ul>

<h1><a name="ScenariotoUseUserScopedCredentialsin%22withCredentials%28%29%22"></a>Scenario to Use User Scoped Credentials in "withCredentials()"</h1>

<p>Many organizations use GPG Signing Key and special permissions on Nexus / Artifactory to create releases. For traceability and security, these privileged credentials may be managed as are "per individual/personal credentials", they may not be shared with team members.</p>

<p>For this kind of credentials, we want to use Jenkins User Scoped Credentials in pipeline (withCredentials, git, config-file-provider, ssh-agent...)</p>

<h1><a name="Description"></a>Description</h1>

<p>When using the authorize project plugin,</p>
<ul>
	<li>User Scoped Credentials are not found by the "withCredentials" pipeline step.</li>
	<li>Global Credentials overwritten by user scoped credentials are not overwritten when used with the "withCredentials" pipeline step.</li>
</ul>

<h1><a name="Reproduce"></a>Reproduce</h1>

<ul>
	<li>Install the Project Authorize Plugin and configure it "Run as user who triggered the build"</li>
	<li>Create a global credential "global-credentials-intended-to-be-overwritten-at-the-user-scope"</li>
	<li>Create user scoped credentials "global-credentials-intended-to-be-overwritten-at-the-user-scope"</li>
	<li>create a pipeline with "withCredentials" binding 'global-bitbucket-credentials-intended-to-be-overwritten-at-the-user-scope' and writing it in a text file</li>
	<li>run the build, open the text file in the workspace and verify that the global credentials are NOT overwritten</li>
</ul>

<ul>
	<li>Create user scoped credentials "my-username-password"</li>
	<li>Create a pipeline with "withCredentials" and the "my-username-password" credentials</li>
	<li>job will fail with "CredentialNotFoundException"</li>
</ul>

<pre>
node {
    <span class="code-comment">// verify that the build is properly impersonated by the https://wiki.jenkins-ci.org/display/JENKINS/Authorize+Project+plugin
</span>    echo <span class="code-quote">"Build is running as user "</span> + org.acegisecurity.context.SecurityContextHolder.getContext().getAuthentication().toString()
    
    stage (<span class="code-quote">"Global Credentials Overwritten at the user scope"</span>) {
<span class="code-comment">// credentials declared globally and overwritten by a user scoped credentials
</span>withCredentials([
    usernamePassword(
credentialsId: <span class="code-quote">'global-credentials-intended-to-be-overwritten-at-the-user-scope'</span>, 
passwordVariable: <span class="code-quote">'PASSWORD_VAR'</span>, 
usernameVariable: <span class="code-quote">'USERNAME_VAR'</span>)]) {
   sh <span class="code-quote">"echo $PASSWORD_VAR > spy-overwritten-creds.txt"</span>
}
    }
    stage (<span class="code-quote">"User Scoped Credentials"</span>) {
withCredentials([
    usernamePassword(
credentialsId: <span class="code-quote">'my-username-password'</span>, 
passwordVariable: <span class="code-quote">'PASSWORD_VAR'</span>, 
usernameVariable: <span class="code-quote">'USERNAME_VAR'</span>)]) {
   sh <span class="code-quote">"echo $PASSWORD_VAR > spy-user-scoped-credentials.txt"</span>
}
    }
}
</pre>

```
Started by user admin
[Pipeline] node
Running on agent-1 in /home/ubuntu/jenkins-aws-home/workspace/tests/user-scoped-credentials-pipeline-step-with-credentials
[Pipeline] {
[Pipeline] echo
Build is running as user org.acegisecurity.providers.UsernamePasswordAuthenticationToken@​965748a4: Username: admin; Password: [PROTECTED]; Authenticated: true; Details: null; Granted Authorities: authenticated
[Pipeline] stage
[Pipeline] { (Global Credentials Overwritten at the user scope)
[Pipeline] withCredentials
[Pipeline] {
[Pipeline] sh
[user-scoped-credentials-pipeline-step-with-credentials] Running shell script
+ echo ****
[Pipeline] }
[Pipeline] // withCredentials
[Pipeline] }
[Pipeline] // stage
[Pipeline] stage
[Pipeline] { (User Scoped Credentials)
[Pipeline] withCredentials
[Pipeline] // withCredentials
[Pipeline] }
[Pipeline] // stage
[Pipeline] }
[Pipeline] // node
[Pipeline] End of Pipeline
org.jenkinsci.plugins.credentialsbinding.impl.CredentialNotFoundException: my-username-password
  at org.jenkinsci.plugins.credentialsbinding.MultiBinding.getCredentials(MultiBinding.java:153)
  at org.jenkinsci.plugins.credentialsbinding.impl.UsernamePasswordMultiBinding.bind(UsernamePasswordMultiBinding.java:76)
  at org.jenkinsci.plugins.credentialsbinding.impl.BindingStep$Execution.start(BindingStep.java:114)
  at org.jenkinsci.plugins.workflow.cps.DSL.invokeStep(DSL.java:184)
  at org.jenkinsci.plugins.workflow.cps.DSL.invokeMethod(DSL.java:126)
  at org.jenkinsci.plugins.workflow.cps.CpsScript.invokeMethod(CpsScript.java:108)
  at groovy.lang.GroovyObject$invokeMethod.call(Unknown Source)
  at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCall(CallSiteArray.java:48)
  at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:113)
  at org.kohsuke.groovy.sandbox.impl.Checker$1.call(Checker.java:151)
  at org.kohsuke.groovy.sandbox.GroovyInterceptor.onMethodCall(GroovyInterceptor.java:21)
  at org.jenkinsci.plugins.scriptsecurity.sandbox.groovy.SandboxInterceptor.onMethodCall(SandboxInterceptor.java:115)
  at org.kohsuke.groovy.sandbox.impl.Checker$1.call(Checker.java:149)
  at org.kohsuke.groovy.sandbox.impl.Checker.checkedCall(Checker.java:146)
  at org.kohsuke.groovy.sandbox.impl.Checker.checkedCall(Checker.java:123)
  at org.kohsuke.groovy.sandbox.impl.Checker.checkedCall(Checker.java:123)
  at com.cloudbees.groovy.cps.sandbox.SandboxInvoker.methodCall(SandboxInvoker.java:16)
  at WorkflowScript.run(WorkflowScript:16)
  at ___cps.transform___(Native Method)
  at com.cloudbees.groovy.cps.impl.ContinuationGroup.methodCall(ContinuationGroup.java:57)
  at com.cloudbees.groovy.cps.impl.FunctionCallBlock$ContinuationImpl.dispatchOrArg(FunctionCallBlock.java:109)
  at com.cloudbees.groovy.cps.impl.FunctionCallBlock$ContinuationImpl.fixArg(FunctionCallBlock.java:82)
  at sun.reflect.GeneratedMethodAccessor591.invoke(Unknown Source)
  at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
  at java.lang.reflect.Method.invoke(Method.java:498)
  at com.cloudbees.groovy.cps.impl.ContinuationPtr$ContinuationImpl.receive(ContinuationPtr.java:72)
  at com.cloudbees.groovy.cps.impl.ClosureBlock.eval(ClosureBlock.java:46)
  at com.cloudbees.groovy.cps.Next.step(Next.java:74)
  at com.cloudbees.groovy.cps.Continuable.run0(Continuable.java:154)
  at org.jenkinsci.plugins.workflow.cps.SandboxContinuable.access$001(SandboxContinuable.java:18)
  at org.jenkinsci.plugins.workflow.cps.SandboxContinuable$1.call(SandboxContinuable.java:33)
  at org.jenkinsci.plugins.workflow.cps.SandboxContinuable$1.call(SandboxContinuable.java:30)
  at org.jenkinsci.plugins.scriptsecurity.sandbox.groovy.GroovySandbox.runInSandbox(GroovySandbox.java:108)
  at org.jenkinsci.plugins.workflow.cps.SandboxContinuable.run0(SandboxContinuable.java:30)
  at org.jenkinsci.plugins.workflow.cps.CpsThread.runNextChunk(CpsThread.java:165)
  at org.jenkinsci.plugins.workflow.cps.CpsThreadGroup.run(CpsThreadGroup.java:328)
  at org.jenkinsci.plugins.workflow.cps.CpsThreadGroup.access$100(CpsThreadGroup.java:80)
  at org.jenkinsci.plugins.workflow.cps.CpsThreadGroup$2.call(CpsThreadGroup.java:240)
  at org.jenkinsci.plugins.workflow.cps.CpsThreadGroup$2.call(CpsThreadGroup.java:228)
  at org.jenkinsci.plugins.workflow.cps.CpsVmExecutorService$2.call(CpsVmExecutorService.java:64)
  at java.util.concurrent.FutureTask.run(FutureTask.java:266)
  at hudson.remoting.SingleLaneExecutorService$1.run(SingleLaneExecutorService.java:112)
  at jenkins.util.ContextResettingExecutorService$1.run(ContextResettingExecutorService.java:28)
  at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
  at java.util.concurrent.FutureTask.run(FutureTask.java:266)
  at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
  at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
  at java.lang.Thread.run(Thread.java:745)
Finished: FAILURE

```

---
<details><summary><i>Originally reported by <img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/cleclerc.png" title="cleclerc's avatar" /> <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=cleclerc">cleclerc</a>, imported from: <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-44772" target="_blank">User Scoped credentials are not used by the "withCredentials" pipeline step</a></i></summary>
<i><ul>
<li><b>assignee</b>: <img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/pskopek.jpg" title="pskopek's avatar" /> <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=pskopek">pskopek</a>
<li><b>status</b>: Open
<li><b>priority</b>: Major
<li><b>component(s)</b>: credentials-binding-plugin
<li><b>resolution</b>: Unresolved
<li><b>votes</b>: 21
<li><b>watchers</b>: 26
<li><b>imported</b>: 20251211-141027
</ul></i>
<details><summary>Raw content of original issue</summary>

<pre>
<h1><a name="Environment"></a>Environment</h1>

<ul>
	<li>Jenkins 2.46.2</li>
	<li>credentials:2.1.13 'Credentials Plugin'</li>
	<li>credentials-binding:1.11 'Credentials Binding Plugin'</li>
	<li>workflow-aggregator:2.5 'Pipeline'</li>
	<li>workflow-api:2.12 'Pipeline: API'</li>
	<li>workflow-basic-steps:2.4 'Pipeline: Basic Steps'</li>
	<li>workflow-cps:2.29 'Pipeline: Groovy'</li>
	<li>workflow-cps-checkpoint:2.4 'CloudBees Pipeline: Groovy Checkpoint Plugin'</li>
	<li>workflow-cps-global-lib:2.7 'Pipeline: Shared Groovy Libraries'</li>
	<li>workflow-durable-task-step:2.10 'Pipeline: Nodes and Processes'</li>
	<li>workflow-job:2.10 'Pipeline: Job'</li>
	<li>workflow-multibranch:2.14 'Pipeline: Multibranch'</li>
	<li>workflow-scm-step:2.4 'Pipeline: SCM Step'</li>
	<li>workflow-step-api:2.9 'Pipeline: Step API'</li>
	<li>workflow-support:2.14 'Pipeline: Supporting APIs'</li>
</ul>

<h1><a name="ScenariotoUseUserScopedCredentialsin%22withCredentials%28%29%22"></a>Scenario to Use User Scoped Credentials in "withCredentials()"</h1>

<p>Many organizations use GPG Signing Key and special permissions on Nexus / Artifactory to create releases. For traceability and security, these privileged credentials may be managed as are "per individual/personal credentials", they may not be shared with team members.</p>

<p>For this kind of credentials, we want to use Jenkins User Scoped Credentials in pipeline (withCredentials, git, config-file-provider, ssh-agent...)</p>

<h1><a name="Description"></a>Description</h1>

<p>When using the authorize project plugin,</p>
<ul>
	<li>User Scoped Credentials are not found by the "withCredentials" pipeline step.</li>
	<li>Global Credentials overwritten by user scoped credentials are not overwritten when used with the "withCredentials" pipeline step.</li>
</ul>

<h1><a name="Reproduce"></a>Reproduce</h1>

<ul>
	<li>Install the Project Authorize Plugin and configure it "Run as user who triggered the build"</li>
	<li>Create a global credential "global-credentials-intended-to-be-overwritten-at-the-user-scope"</li>
	<li>Create user scoped credentials "global-credentials-intended-to-be-overwritten-at-the-user-scope"</li>
	<li>create a pipeline with "withCredentials" binding 'global-bitbucket-credentials-intended-to-be-overwritten-at-the-user-scope' and writing it in a text file</li>
	<li>run the build, open the text file in the workspace and verify that the global credentials are NOT overwritten</li>
</ul>

<ul>
	<li>Create user scoped credentials "my-username-password"</li>
	<li>Create a pipeline with "withCredentials" and the "my-username-password" credentials</li>
	<li>job will fail with "CredentialNotFoundException"</li>
</ul>

<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">node {
    <span class="code-comment">// verify that the build is properly impersonated by the https://wiki.jenkins-ci.org/display/JENKINS/Authorize+Project+plugin
</span>    echo <span class="code-quote">"Build is running as user "</span> + org.acegisecurity.context.SecurityContextHolder.getContext().getAuthentication().toString()
    
    stage (<span class="code-quote">"Global Credentials Overwritten at the user scope"</span>) {
        <span class="code-comment">// credentials declared globally and overwritten by a user scoped credentials
</span>        withCredentials([
            usernamePassword(
                credentialsId: <span class="code-quote">'global-credentials-intended-to-be-overwritten-at-the-user-scope'</span>, 
                passwordVariable: <span class="code-quote">'PASSWORD_VAR'</span>, 
                usernameVariable: <span class="code-quote">'USERNAME_VAR'</span>)]) {
           sh <span class="code-quote">"echo $PASSWORD_VAR &gt; spy-overwritten-creds.txt"</span>
        }
    }
    stage (<span class="code-quote">"User Scoped Credentials"</span>) {
        withCredentials([
            usernamePassword(
                credentialsId: <span class="code-quote">'my-username-password'</span>, 
                passwordVariable: <span class="code-quote">'PASSWORD_VAR'</span>, 
                usernameVariable: <span class="code-quote">'USERNAME_VAR'</span>)]) {
           sh <span class="code-quote">"echo $PASSWORD_VAR &gt; spy-user-scoped-credentials.txt"</span>
        }
    }
}
</pre>
</div></div>

<div class="preformatted panel" style="border-width: 1px;"><div class="preformattedContent panelContent">
<pre>Started by user admin
[Pipeline] node
Running on agent-1 in /home/ubuntu/jenkins-aws-home/workspace/tests/user-scoped-credentials-pipeline-step-with-credentials
[Pipeline] {
[Pipeline] echo
Build is running as user org.acegisecurity.providers.UsernamePasswordAuthenticationToken@965748a4: Username: admin; Password: [PROTECTED]; Authenticated: true; Details: null; Granted Authorities: authenticated
[Pipeline] stage
[Pipeline] { (Global Credentials Overwritten at the user scope)
[Pipeline] withCredentials
[Pipeline] {
[Pipeline] sh
[user-scoped-credentials-pipeline-step-with-credentials] Running shell script
+ echo ****
[Pipeline] }
[Pipeline] // withCredentials
[Pipeline] }
[Pipeline] // stage
[Pipeline] stage
[Pipeline] { (User Scoped Credentials)
[Pipeline] withCredentials
[Pipeline] // withCredentials
[Pipeline] }
[Pipeline] // stage
[Pipeline] }
[Pipeline] // node
[Pipeline] End of Pipeline
org.jenkinsci.plugins.credentialsbinding.impl.CredentialNotFoundException: my-username-password
  at org.jenkinsci.plugins.credentialsbinding.MultiBinding.getCredentials(MultiBinding.java:153)
  at org.jenkinsci.plugins.credentialsbinding.impl.UsernamePasswordMultiBinding.bind(UsernamePasswordMultiBinding.java:76)
  at org.jenkinsci.plugins.credentialsbinding.impl.BindingStep$Execution.start(BindingStep.java:114)
  at org.jenkinsci.plugins.workflow.cps.DSL.invokeStep(DSL.java:184)
  at org.jenkinsci.plugins.workflow.cps.DSL.invokeMethod(DSL.java:126)
  at org.jenkinsci.plugins.workflow.cps.CpsScript.invokeMethod(CpsScript.java:108)
  at groovy.lang.GroovyObject$invokeMethod.call(Unknown Source)
  at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCall(CallSiteArray.java:48)
  at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:113)
  at org.kohsuke.groovy.sandbox.impl.Checker$1.call(Checker.java:151)
  at org.kohsuke.groovy.sandbox.GroovyInterceptor.onMethodCall(GroovyInterceptor.java:21)
  at org.jenkinsci.plugins.scriptsecurity.sandbox.groovy.SandboxInterceptor.onMethodCall(SandboxInterceptor.java:115)
  at org.kohsuke.groovy.sandbox.impl.Checker$1.call(Checker.java:149)
  at org.kohsuke.groovy.sandbox.impl.Checker.checkedCall(Checker.java:146)
  at org.kohsuke.groovy.sandbox.impl.Checker.checkedCall(Checker.java:123)
  at org.kohsuke.groovy.sandbox.impl.Checker.checkedCall(Checker.java:123)
  at com.cloudbees.groovy.cps.sandbox.SandboxInvoker.methodCall(SandboxInvoker.java:16)
  at WorkflowScript.run(WorkflowScript:16)
  at ___cps.transform___(Native Method)
  at com.cloudbees.groovy.cps.impl.ContinuationGroup.methodCall(ContinuationGroup.java:57)
  at com.cloudbees.groovy.cps.impl.FunctionCallBlock$ContinuationImpl.dispatchOrArg(FunctionCallBlock.java:109)
  at com.cloudbees.groovy.cps.impl.FunctionCallBlock$ContinuationImpl.fixArg(FunctionCallBlock.java:82)
  at sun.reflect.GeneratedMethodAccessor591.invoke(Unknown Source)
  at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
  at java.lang.reflect.Method.invoke(Method.java:498)
  at com.cloudbees.groovy.cps.impl.ContinuationPtr$ContinuationImpl.receive(ContinuationPtr.java:72)
  at com.cloudbees.groovy.cps.impl.ClosureBlock.eval(ClosureBlock.java:46)
  at com.cloudbees.groovy.cps.Next.step(Next.java:74)
  at com.cloudbees.groovy.cps.Continuable.run0(Continuable.java:154)
  at org.jenkinsci.plugins.workflow.cps.SandboxContinuable.access$001(SandboxContinuable.java:18)
  at org.jenkinsci.plugins.workflow.cps.SandboxContinuable$1.call(SandboxContinuable.java:33)
  at org.jenkinsci.plugins.workflow.cps.SandboxContinuable$1.call(SandboxContinuable.java:30)
  at org.jenkinsci.plugins.scriptsecurity.sandbox.groovy.GroovySandbox.runInSandbox(GroovySandbox.java:108)
  at org.jenkinsci.plugins.workflow.cps.SandboxContinuable.run0(SandboxContinuable.java:30)
  at org.jenkinsci.plugins.workflow.cps.CpsThread.runNextChunk(CpsThread.java:165)
  at org.jenkinsci.plugins.workflow.cps.CpsThreadGroup.run(CpsThreadGroup.java:328)
  at org.jenkinsci.plugins.workflow.cps.CpsThreadGroup.access$100(CpsThreadGroup.java:80)
  at org.jenkinsci.plugins.workflow.cps.CpsThreadGroup$2.call(CpsThreadGroup.java:240)
  at org.jenkinsci.plugins.workflow.cps.CpsThreadGroup$2.call(CpsThreadGroup.java:228)
  at org.jenkinsci.plugins.workflow.cps.CpsVmExecutorService$2.call(CpsVmExecutorService.java:64)
  at java.util.concurrent.FutureTask.run(FutureTask.java:266)
  at hudson.remoting.SingleLaneExecutorService$1.run(SingleLaneExecutorService.java:112)
  at jenkins.util.ContextResettingExecutorService$1.run(ContextResettingExecutorService.java:28)
  at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
  at java.util.concurrent.FutureTask.run(FutureTask.java:266)
  at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
  at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
  at java.lang.Thread.run(Thread.java:745)
Finished: FAILURE
</pre>
</div></div></pre>
</details>
</details>
<details><summary><i>1 attachment</i></summary>

- [cloudbees-support_cje-master.alliances.beescloud.com_2017-04-27_14.09.05.zip](https://issues.jenkins.io/secure/attachment/38405/cloudbees-support_cje-master.alliances.beescloud.com_2017-04-27_14.09.05.zip)
</details>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/matthiasb.svg" title="matthiasb's avatar" /> matthiasb:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-44772?focusedId=350562&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-350562">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Thanks <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=pskopek" class="user-hover" rel="pskopek">pskopek</a> for this PR, what is the actual state of this issue/PR?</p>

<p>We have a similar problem, because we run deployments based to the user scoped credentials and this don't work since a while.
So we would like to see this PR merged. <img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/smile.png" height="16" width="16" align="absmiddle" alt="" border="0"/></p>
</pre>
</details>

</ul>
</details>

<p>Thanks <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=pskopek" class="user-hover" rel="pskopek">pskopek</a> for this PR, what is the actual state of this issue/PR?</p>

<p>We have a similar problem, because we run deployments based to the user scoped credentials and this don't work since a while.<br/>
So we would like to see this PR merged. <img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/smile.png" height="16" width="16" align="absmiddle" alt="" border="0"/></p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/fernando_rosado.png" title="fernando_rosado's avatar" /> fernando_rosado:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-44772?focusedId=374769&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-374769">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>I will clarify, i could use user scope credential using the workarround commented on following issue ticket : 
<a href="https://issues.jenkins-ci.org/browse/JENKINS-38963?focusedCommentId=318189&amp;page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-318189" class="external-link" target="_blank" rel="nofollow noopener">https://issues.jenkins-ci.org/browse/JENKINS-38963?focusedCommentId=318189&amp;page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-318189</a> </p>

<p>The use of the single quotes around '${credential_id}' is mandatory, it is not the same issue commented (overwrite a credentialID) but perhaps exists some relation. </p>
</pre>
</details>

</ul>
</details>

<p>I will clarify, i could use user scope credential using the workarround commented on following issue ticket : <br/>
<a href="https://issues.jenkins-ci.org/browse/JENKINS-38963?focusedCommentId=318189&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-318189" class="external-link" target="_blank" rel="nofollow noopener">https://issues.jenkins-ci.org/browse/JENKINS-38963?focusedCommentId=318189&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-318189</a> </p>

<p>The use of the single quotes around '${credential_id}' is mandatory, it is not the same issue commented (overwrite a credentialID) but perhaps exists some relation. </p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>vit_zikmund:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-44772?focusedId=391934&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-391934">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Hello <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jvz" class="user-hover" rel="jvz">jvz</a>, I'm quite sure it is not the case. Your JENIKINS-58170 is nice, but it doesn't allow you to access the userscope credentials without them being supplied in a parameter.
I suppose <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=cleclerc" class="user-hover" rel="cleclerc">cleclerc</a> was trying to avoid using the build parameter entirely.</p>

<p>I personally don't see the security gain in this, especially if it's me, who's running the job, why wouldn't I want the job to access my user credentials? I know there's a possibility that if this got enabled, the job could guess and snoop any user's credential, but there could also be a way for the job to declare which credentials it's about to be accessing which the user would (until further change) accept only once and use it without selecting a credentials parameter each time s/he runs the build.</p>

<p>I actually found a way how to circumvent the limitation by creating an upstream job, that uses a user's jenkins API token to trigger the job with credentials via the Jenkins API (so the downstream build thinks it was <em>directly</em> run by the user, which is one of the requirements for the user creds to be read), but it required me/my peers to reinvent the whole build() logic with curl, and that just sucked. I'd still give an eternal praise to someone willing to reconsider these limitations to make them more user friendly.</p>
</pre>
</details>

</ul>
</details>

<p>Hello <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jvz" class="user-hover" rel="jvz">jvz</a>, I'm quite sure it is not the case. Your JENIKINS-58170 is nice, but it doesn't allow you to access the userscope credentials without them being supplied in a parameter.<br/>
I suppose <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=cleclerc" class="user-hover" rel="cleclerc">cleclerc</a> was trying to avoid using the build parameter entirely.</p>

<p>I personally don't see the security gain in this, especially if it's me, who's running the job, why wouldn't I want the job to access my user credentials? I know there's a possibility that if this got enabled, the job could guess and snoop any user's credential, but there could also be a way for the job to declare which credentials it's about to be accessing which the user would (until further change) accept only once and use it without selecting a credentials parameter each time s/he runs the build.</p>

<p>I actually found a way how to circumvent the limitation by creating an upstream job, that uses a user's jenkins API token to trigger the job with credentials via the Jenkins API (so the downstream build thinks it was <em>directly</em> run by the user, which is one of the requirements for the user creds to be read), but it required me/my peers to reinvent the whole build() logic with curl, and that just sucked. I'd still give an eternal praise to someone willing to reconsider these limitations to make them more user friendly.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>vit_zikmund:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-44772?focusedId=402366&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-402366">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Hi <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jvz" class="user-hover" rel="jvz">jvz</a>, here's what we deal with. First of all, we do use the Authorize Project Plugin with all jobs set to: <tt>Run as the user who triggered the build.</tt></p>

<p>There's one "master" job with a credential parameter. Via that one, we can access the selected user-scope credential and all is nice and smooth.</p>

<p>The problem is, that this job triggers another builds via the <a href="https://www.jenkins.io/doc/pipeline/steps/pipeline-build-step/" class="external-link" target="_blank" rel="nofollow noopener"><tt>build()</tt> step</a> and those builds need the "master" build credential as well. And that's where we hit the wall, or mud at least, as there's an ugly way through, but the credentials parameter is not a big help in either of the workarounds we've figured out:</p>
<ol>
	<li>Use the credentials parameter in the downstream jobs and trigger them via Jenkins API and the user's auth token to pretend it was directly him/her who did it.
 This allows the authorize/credentials plugins to work properly and reach for the user's creds.
 The downside is that the triggering from the "master" job can't be done with the <tt>build()</tt> step, as that puts the master build as the main trigger cause, which IMHO kills the authorize plugin's <tt>Run as the user who triggered the build</tt> (even though the original user-triggered cause is also transitively visible)
 In one case we really reinvented the <tt>build()</tt> wheel to do that, and we made the users to save their auth tokens in their user creds and trigger the "master" build with it.
 For illustration the core looks like this:
 <tt>curl -</tt><tt>sS -f -i -X POST "https://$JENKINS_HOSTNAME/job/$JOBNAME/buildWithParameters" --user "$USERNAME:$TOKEN" </tt><tt>-data-urlencode "user_cred=$DESIRED_USER_CRED_ID"</tt>
 Sad things:</li>
</ol>


<ul>
	<li>rewritten <tt>build()</tt> in "user space"</li>
	<li>user has to manage one more credential than necessary (the auth token)</li>
	<li>the overall security of this is a bit questionable.</li>
</ul>


<ol>
	<li>Pass the credential from the "master" build downstream as a Password parameter.
 That's pretty straightforward, we extract the desired credential in the "master" build and pass it as a string downstream via a password parameter. (recently we had fun with <a href="https://issues.jenkins.io/browse/JENKINS-62305" title="Error &quot;hudson.model.PasswordParameterValue.value expects class hudson.util.Secret&quot; since 2.236" class="issue-link" data-issue-key="JENKINS-62305"><del>JENKINS-62305</del></a>, but that's another story).
 Downsides:</li>
</ol>


<ul>
	<li>If the "master" build doesn't need the user credential, it has to access it anyways to pass it downstream.</li>
	<li>Duplicate parameters for the same thing - credential for those who run the downstream job directly, password for the "master" job.</li>
	<li>To at least hide this from the Build With Parameters screen, I've been omitting the password parameter declaration, but without it jenkins floods the logs with a security error about that it's being passed an undeclared parameter.</li>
</ul>


<p>Ideally, the downstream job would be able to access the user creds in both cases:</p>
<ul>
	<li>being triggered directly by the user</li>
	<li>being triggered by an upstream build triggered by the user.</li>
</ul>


<p>If there's a way how to do that and have only one code / creds parameter for the downstream job and just passing the cred id from the upstream/"master" job, that would be a big relief. Right now, I don't see that option, but I'm not yet sure what the do next.</p>

<p>You also say</p>
<blockquote><p>You can attach the actions directly via the low level API to simulate credential authorization by a user.</p></blockquote>
<p>Can you elaborate on that a bit, toss pointers to some docs or examples if that's still relevant?</p>

<p>Thank you very much and sorry for the long post!</p>
</pre>
</details>

</ul>
</details>

<p>Hi <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jvz" class="user-hover" rel="jvz">jvz</a>, here's what we deal with. First of all, we do use the Authorize Project Plugin with all jobs set to: <tt>Run as the user who triggered the build.</tt></p>

<p>There's one "master" job with a credential parameter. Via that one, we can access the selected user-scope credential and all is nice and smooth.</p>

<p>The problem is, that this job triggers another builds via the <a href="https://www.jenkins.io/doc/pipeline/steps/pipeline-build-step/" class="external-link" target="_blank" rel="nofollow noopener"><tt>build()</tt> step</a> and those builds need the "master" build credential as well. And that's where we hit the wall, or mud at least, as there's an ugly way through, but the credentials parameter is not a big help in either of the workarounds we've figured out:</p>
<ol>
	<li>Use the credentials parameter in the downstream jobs and trigger them via Jenkins API and the user's auth token to pretend it was directly him/her who did it.<br/>
 This allows the authorize/credentials plugins to work properly and reach for the user's creds.<br/>
 The downside is that the triggering from the "master" job can't be done with the <tt>build()</tt> step, as that puts the master build as the main trigger cause, which IMHO kills the authorize plugin's <tt>Run as the user who triggered the build</tt> (even though the original user-triggered cause is also transitively visible)<br/>
 In one case we really reinvented the <tt>build()</tt> wheel to do that, and we made the users to save their auth tokens in their user creds and trigger the "master" build with it.<br/>
 For illustration the core looks like this:<br/>
 <tt>curl -</tt><tt>sS -f -i -X POST "https://$JENKINS_HOSTNAME/job/$JOBNAME/buildWithParameters" --user "$USERNAME:$TOKEN" </tt><tt>-data-urlencode "user_cred=$DESIRED_USER_CRED_ID"</tt><br/>
 Sad things:</li>
</ol>


<ul>
	<li>rewritten <tt>build()</tt> in "user space"</li>
	<li>user has to manage one more credential than necessary (the auth token)</li>
	<li>the overall security of this is a bit questionable.</li>
</ul>


<ol>
	<li>Pass the credential from the "master" build downstream as a Password parameter.<br/>
 That's pretty straightforward, we extract the desired credential in the "master" build and pass it as a string downstream via a password parameter. (recently we had fun with <a href="https://issue-redirect.jenkins.io/issue/62305" title="Error "hudson.model.PasswordParameterValue.value expects class hudson.util.Secret" since 2.236" class="issue-link" data-issue-key="<a class="jira-plain-text-key" href="https://issue-redirect.jenkins.io/issue/62305">JENKINS-62305</a>"><del>JENKINS-62305</del></a>, but that's another story).<br/>
 Downsides:</li>
</ol>


<ul>
	<li>If the "master" build doesn't need the user credential, it has to access it anyways to pass it downstream.</li>
	<li>Duplicate parameters for the same thing - credential for those who run the downstream job directly, password for the "master" job.</li>
	<li>To at least hide this from the Build With Parameters screen, I've been omitting the password parameter declaration, but without it jenkins floods the logs with a security error about that it's being passed an undeclared parameter.</li>
</ul>


<p>Ideally, the downstream job would be able to access the user creds in both cases:</p>
<ul>
	<li>being triggered directly by the user</li>
	<li>being triggered by an upstream build triggered by the user.</li>
</ul>


<p>If there's a way how to do that and have only one code / creds parameter for the downstream job and just passing the cred id from the upstream/"master" job, that would be a big relief. Right now, I don't see that option, but I'm not yet sure what the do next.</p>

<p>You also say</p>
<blockquote><p>You can attach the actions directly via the low level API to simulate credential authorization by a user.</p></blockquote>
<p>Can you elaborate on that a bit, toss pointers to some docs or examples if that's still relevant?</p>

<p>Thank you very much and sorry for the long post!</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/jvz.png" title="jvz's avatar" /> jvz:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-44772?focusedId=402373&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-402373">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>You've primarily described <a href="https://issues.jenkins.io/browse/JENKINS-59109" class="external-link" rel="nofollow">https://issues.jenkins.io/browse/JENKINS-59109</a> which would add support for propagating user credentials from one build to another via the build step. The low level API is described here: <a href="https://github.com/jenkinsci/credentials-plugin/blob/master/docs/consumer.adoc#binding-user-supplied-credentials-parameters-to-builds" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/jenkinsci/credentials-plugin/blob/master/docs/consumer.adoc#binding-user-supplied-credentials-parameters-to-builds</a> which would be used in an implementation for <a href="https://issues.jenkins.io/browse/JENKINS-59109" title="Allow user scoped credentials to be used by build step" class="issue-link" data-issue-key="JENKINS-59109">JENKINS-59109</a> and any other plugins that wish to integrate this further. Ideally, this feature would offer the ability to specify which loaded credentials to propagate. The gist of the integration would propagate the user who bound each credential parameter so that downstream jobs still know which user corresponds to which credential binding for authorization purposes.</p>
</pre>
</details>

</ul>
</details>

<p>You've primarily described <a href="https://issue-redirect.jenkins.io/issue/59109" class="external-link" rel="nofollow">https://issue-redirect.jenkins.io/issue/59109</a> which would add support for propagating user credentials from one build to another via the build step. The low level API is described here: <a href="https://github.com/jenkinsci/credentials-plugin/blob/master/docs/consumer.adoc#binding-user-supplied-credentials-parameters-to-builds" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/jenkinsci/credentials-plugin/blob/master/docs/consumer.adoc#binding-user-supplied-credentials-parameters-to-builds</a> which would be used in an implementation for <a href="https://issue-redirect.jenkins.io/issue/59109" title="Allow user scoped credentials to be used by build step" class="issue-link" data-issue-key="<a class="jira-plain-text-key" href="https://issue-redirect.jenkins.io/issue/59109">JENKINS-59109</a>">JENKINS-59109</a> and any other plugins that wish to integrate this further. Ideally, this feature would offer the ability to specify which loaded credentials to propagate. The gist of the integration would propagate the user who bound each credential parameter so that downstream jobs still know which user corresponds to which credential binding for authorization purposes.</p>

**Comment by jenkins-infra-bot** [comment]
<i>[Original `is_duplicated_by` from Jira: <a class="jira-relationship" href="https://issue-redirect.jenkins.io/issue/44635">JENKINS-44635</a>]</i>

**Comment by jenkins-infra-bot** [comment]
<i>[Original `relates_to` from Jira: <a class="jira-relationship" href="https://issue-redirect.jenkins.io/issue/55052">JENKINS-55052</a>]</i>

**Comment by jenkins-infra-bot** [comment]
<i>[Original `relates_to` from Jira: <a class="jira-relationship" href="https://issue-redirect.jenkins.io/issue/38963">JENKINS-38963</a>]</i>

**Comment by jenkins-infra-bot** [comment]
<i>[Original `relates_to` from Jira: <a class="jira-relationship" href="https://issue-redirect.jenkins.io/issue/44774">JENKINS-44774</a>]</i>
